### PR TITLE
Refactor of TIAF Code

### PR DESCRIPTION
**Pull request by jcurtisk**
## What does this PR do?
This PR is a refactor of the main TIAF script, to hopefully make it more extensible in the future while retaining functionality. A minor bug has been fixed within this PR, meaning that the historic data flag is now correctly set on generated reports.

Note: Diff is a little confusing, a lot of the code has been migrated from the `run` method into the `__init__` method. 
## How was this PR tested?
Manually tested by inspection and comparing console output and output to reporting software(kibana)

### CODE REVIEW

File: scripts/build/TestImpactAnalysis/tiaf.py [hunk 1]
```diff
@@ -18,16 +18,227 @@
 
 logger = get_logger(__file__)
 
+
 class TestImpact:
-    def __init__(self, config_file: str):
+
+    
+
```
Comment (hultonha) [inline scripts/build/TestImpactAnalysis/tiaf.py:25]: uber nit: Are all these line breaks necessary (I think python recommends two between statements/functions is that correct?

File: scripts/build/TestImpactAnalysis/tiaf.py [hunk 2]
```diff
@@ -18,16 +18,227 @@
...
+        self._s3_bucket = args.get('s3_bucket')
+        self._suite = args.get('suite')
+
+        self._config = self._parse_config_file(args['config'])
+
+        # Initialise branches
+        self._src_branch = args.get("src_branch")
+        self._dst_branch = args.get("dst_branch")
+        logger.info(f"Src branch: '{self._src_branch}'.")
+        logger.info(f"Dst branch: '{self._dst_branch}'.")
```
Comment (hultonha) [inline scripts/build/TestImpactAnalysis/tiaf.py:51]: nit: In the logging output please prefer full names, so use 'Source branch' and 'Destination branch' for clarity

File: scripts/build/TestImpactAnalysis/tiaf.py [hunk 3]
```diff
@@ -18,16 +18,227 @@
...
+
+        # Determine our source of truth. Also initialises our source of truth property.
+        self._determine_source_of_truth()
+
+        # Initialise commit info
```
Comment (hultonha) [inline scripts/build/TestImpactAnalysis/tiaf.py:56]: uber nit (only as I've been burned by this a few times 😅) please use 'American' English for all code (purely for consistency). So this is 'initializes' (more important for code names but good good comments too)

File: scripts/build/TestImpactAnalysis/tiaf.py [hunk 4]
```diff
@@ -18,16 +18,227 @@
...
+        self._dst_commit = args.get("commit")
+        logger.info(f"Commit: '{self._dst_commit}'.")
+        self._src_commit = None
```
Comment (hultonha) [inline scripts/build/TestImpactAnalysis/tiaf.py:59]: nit: It's probably okay here, but please use abbreviations sparingly, code is read far more often than it is written so having slightly longer but more readable names is nearly always the right tradeoff 🙂

File: scripts/build/TestImpactAnalysis/tiaf.py [hunk 5]
```diff
@@ -18,16 +18,227 @@
...
+                    logger.info("No historic data found.")
+
+                # Determining our sequence type:
+                if self._has_change_list:
+                    if self._is_source_of_truth_branch:
+                        # Use TIA sequence (instrumented subset of tests) for coverage updating branches so we can update the coverage data with the generated coverage
+                        sequence_type = "tia"
+                    else:
+                        # Use TIA no-write sequence (regular subset of tests) for non coverage updating branche
+                        sequence_type = "tianowrite"
```
Comment (hultonha) [inline scripts/build/TestImpactAnalysis/tiaf.py:90]: question: Could this be tia_no_write instead to make it a little easier to parse?

File: scripts/build/TestImpactAnalysis/tiaf.py [hunk 6]
```diff
@@ -18,16 +18,227 @@
...
+                        # Ignore integrity failures for non coverage updating branches as our confidence in the
```
Comment (hultonha) [inline scripts/build/TestImpactAnalysis/tiaf.py:91]: question: This sentence seems to end abruptly 😅

File: scripts/build/TestImpactAnalysis/tiaf.py [hunk 7]
```diff
@@ -18,16 +18,227 @@
...
+                            "Integration failure policy is set to 'continue'.")
+
+        self._parse_arguments_to_runtime(
+            args, sequence_type, self._runtime_args)
+
+    def _parse_arguments_to_runtime(self, args, sequence_type, runtime_args):
+        """
+        Fetches the relevant keys from the provided dict, and applies the values of the arguments(or applies them as a flag) to our runtime_args list.
+
+        @param args: Dict containing the arguments passed to this TestImpact object. Will contain all the runtime arguments we need to apply.
```
Comment (hultonha) [inline scripts/build/TestImpactAnalysis/tiaf.py:118]: nit: Again here in the comment it's not bad to spell out 'Dictionary'

File: scripts/build/TestImpactAnalysis/tiaf.py [hunk 8]
```diff
@@ -18,16 +18,227 @@
...
+        # Test failure policy
+        test_failure_policy = args.get('test_failure_policy')
+        runtime_args.append(f"--fpolicy={test_failure_policy}")
+        logger.info(f"Test failure policy is set to '{test_failure_policy}'.")
+
+        # Sequence report
+        self._report_file = pathlib.PurePath(self._temp_workspace).joinpath(
+            f"report.{self._instance_id}.json")
+        runtime_args.append(f"--report={self._report_file}")
+        logger.info(f"Sequence report file is set to '{self._report_file}'.")
```
Comment (hultonha) [inline scripts/build/TestImpactAnalysis/tiaf.py:141]: All these logs are great 👍

File: scripts/build/TestImpactAnalysis/tiaf.py [hunk 9]
```diff
@@ -18,16 +18,227 @@
...
+                return PersistentStorageLocal(
+                    self._config, suite, self._dst_commit)
+        except SystemError as e:
+            logger.warning(
+                f"The persistent storage encountered an irrecoverable error, test impact analysis will be disabled: '{e}'")
+            return None
+
+    def _determine_source_of_truth(self):
+        """
+        Determines whether the branch we are executing TIAF on is the source of truth(the branch from which the coverage data will be stored/retrieved from) or not.
```
Comment (hultonha) [inline scripts/build/TestImpactAnalysis/tiaf.py:228]: uber nit: add space before paren in the comment

File: scripts/build/TestImpactAnalysis/tiaf_driver.py [hunk 10]
```diff
@@ -142,32 +144,31 @@ def valid_test_failure_policy(value):
...
         logger.error(f"Exception caught by TIAF driver: '{e}'.")
         traceback.print_exc()
+
+
+if __name__ == "__main__":
+
+    args = vars(parse_args())
+    main(args)
+
+
```
Comment (hultonha) [inline scripts/build/TestImpactAnalysis/tiaf_driver.py:174]: uber nit: EOL character missing

File: scripts/build/TestImpactAnalysis/tiaf_persistent_storage.py [hunk 11]
```diff
@@ -163,20 +163,36 @@ def update_and_store_historic_data(self):
...
+        Hash of this commit. Is none if this commit hash was not found in our historic data.
+        """
         return self._this_commit_hash_last_commit_hash
 
     @property
-    def can_rerun_sequence(self):
+    def has_previous_last_commit_hash(self):
+        """
+        If the hash of the last commit was found in our historic data, then this flag will be set.
+        """
```
Comment (hultonha) [inline scripts/build/TestImpactAnalysis/tiaf_persistent_storage.py:197]: These added comments are great, thanks for taking the time to add them :+1:

File: scripts/build/TestImpactAnalysis/tiaf_persistent_storage.py [hunk 12]
```diff
@@ -163,20 +163,36 @@ def update_and_store_historic_data(self):
...
         return self._has_previous_last_commit_hash
```
Comment (hultonha) [inline scripts/build/TestImpactAnalysis/tiaf_persistent_storage.py:198]: uber nit: EOL character missing

**Review by hultonha — APPROVED** [review]
These changes look good to me, likely worth @jonawals going through too. Nice work @jcurtisk!

File: scripts/build/TestImpactAnalysis/tiaf.py (hunk 2 above)
Comment (jcurtisk) [reply scripts/build/TestImpactAnalysis/tiaf.py:51]: Fixed.

File: scripts/build/TestImpactAnalysis/tiaf.py (hunk 5 above)
Comment (jcurtisk) [reply scripts/build/TestImpactAnalysis/tiaf.py:90]: It could be, this could be a useful change but would require modifying the C++ runtime to accept this. @jonawals Thoughts?

File: scripts/build/TestImpactAnalysis/tiaf.py (hunk 6 above)
Comment (jcurtisk) [reply scripts/build/TestImpactAnalysis/tiaf.py:91]: Fixed

File: scripts/build/TestImpactAnalysis/tiaf.py (hunk 4 above)
Comment (jcurtisk) [reply scripts/build/TestImpactAnalysis/tiaf.py:58]: Will keep in mind.

File: scripts/build/TestImpactAnalysis/tiaf.py (hunk 7 above)
Comment (jcurtisk) [reply scripts/build/TestImpactAnalysis/tiaf.py]: Noted, have fixed.

File: scripts/build/TestImpactAnalysis/tiaf.py (hunk 1 above)
Comment (hultonha) [reply scripts/build/TestImpactAnalysis/tiaf.py]: Nice one thanks!

File: scripts/build/TestImpactAnalysis/tiaf.py (hunk 2 above)
Comment (hultonha) [reply scripts/build/TestImpactAnalysis/tiaf.py]: Thanks! 👍

**Review by hultonha — APPROVED** [review]

File: scripts/build/TestImpactAnalysis/tiaf_persistent_storage.py (hunk 12 above)
Comment (jonawals) [reply scripts/build/TestImpactAnalysis/tiaf_persistent_storage.py:198]: @jcurtisk

**Review by jonawals — COMMENTED** [review]
Looks good, it's hard to parse the diff so i will approve on faith now, let's verify it with your unit tests 👍
